File: mcp/src/quarantine/state-machine.ts
// Pure state-transition functions. No I/O, no side effects.
// Single source of truth for what transitions are legal.
//
// Dual-backend: TypeScript by default (V0), or Rust via the @mafia/core-node
// FFI binding when MAFIA_CORE_BACKEND=rust. Same observable behavior; the
// matrix test (`npm run test:matrix`) runs the full suite against both.
//
// State diagram:
//
//   flagged ──quarantine──▶ quarantining ──quarantine_complete──▶ quarantined
//      │                          │                                     │
//      └──keep──▶ kept             └──fail──▶ failed                    │
//                                                                       │
//                                          ┌──restore──▶ restoring ─────┤
//                                          │              │             │
//                                          │              ├─restore_complete──▶ restored
//                                          │              └─fail──▶ failed
//                                          │                            │
//                                          └──purge──▶ purging ◀────────┘
//                                                       │
//                                                       ├─purge_complete──▶ purged
//                                                       └─fail──▶ failed

import type { State, Transition } from './types.js';
import * as rustCore from '@mafia/core-node';

const BACKEND = (process.env.MAFIA_CORE_BACKEND ?? 'typescript').toLowerCase();
const USE_RUST = BACKEND === 'rust';

/** Which backend is active. Exposed for diagnostics + eval reporting. */
export function activeBackend(): 'rust' | 'typescript' {
  return USE_RUST ? 'rust' : 'typescript';
}

interface TransitionRule {
  from: State[];
  to: State;
}

const RULES: Record<Exclude<Transition, 'genesis'>, TransitionRule> = {
  // Initial flagging is a row creation, not a transition between existing
  // states; flag is treated as transition into 'flagged' from a notional
  // 'pre' state.
  flag: { from: [], to: 'flagged' },

  keep: { from: ['flagged'], to: 'kept' },

  quarantine: { from: ['flagged'], to: 'quarantining' },
  quarantine_complete: { from: ['quarantining'], to: 'quarantined' },

  restore: { from: ['quarantined'], to: 'restoring' },
  restore_complete: { from: ['restoring'], to: 'restored' },

  purge: { from: ['quarantined'], to: 'purging' },
  purge_complete: { from: ['purging'], to: 'purged' },

  // Any *-ing state may transition to failed if the upstream call errors.
  fail: { from: ['quarantining', 'restoring', 'purging'], to: 'failed' },
};

export class IllegalTransition extends Error {
  constructor(
    public readonly from: State | null,
    public readonly transition: Transition,
  ) {
    super(`Illegal transition: ${from ?? '<new>'} --${transition}--> ?`);
    this.name = 'IllegalTransition';
  }
}

/**
 * Validate that a transition is legal from the current state.
 * Returns the resulting state, or throws IllegalTransition.
 */
export function nextState(from: State | null, transition: Transition): State {
  return USE_RUST ? nextStateRust(from, transition) : nextStateTs(from, transition);
}

function nextStateTs(from: State | null, transition: Transition): State {
  if (transition === 'genesis') {
    throw new IllegalTransition(from, transition); // genesis is reflog-only
  }
  const rule = RULES[transition];
  if (transition === 'flag') {
    if (from !== null) throw new IllegalTransition(from, transition);
    return rule.to;
  }
  if (from === null) throw new IllegalTransition(from, transition);
  if (!rule.from.includes(from)) throw new IllegalTransition(from, transition);
  return rule.to;
}

function nextStateRust(from: State | null, transition: Transition): State {
  try {
    return rustCore.nextState(from, transition) as State;
  } catch {
    // Re-wrap so callers still see our IllegalTransition class.
    throw new IllegalTransition(from, transition);
  }
}

/**
 * Convenience: is the given state terminal? Terminal states never transition further.
 */
export function isTerminal(state: State): boolean {
  if (USE_RUST) return rustCore.isTerminal(state);
  return state === 'kept' || state === 'restored' || state === 'purged' || state === 'failed';
}

/**
 * Convenience: is the given state in-flight? In-flight states are recovered
 * on startup by the outbox reconcile pass.
 */
export function isInFlight(state: State): boolean {
  if (USE_RUST) return rustCore.isInFlight(state);
  return state === 'quarantining' || state === 'restoring' || state === 'purging';
}

/**
 * Reflog-relevant transitions, in canonical order. Useful for tests.
 */
export const ALL_TRANSITIONS: Transition[] = [
  'flag',
  'keep',
  'quarantine',
  'quarantine_complete',
  'restore',
  'restore_complete',
  'purge',
  'purge_complete',
  'fail',
];

File: core/tests/state_machine.rs
//! Fixture tests mirroring `Mafia/tests/state-machine.test.ts`.
//! These MUST pass with identical semantics on both sides.

use mafia_core::state_machine::{is_in_flight, is_terminal};
use mafia_core::{next_state, IllegalTransition, State, Transition};

// ── Flag ─────────────────────────────────────────────────────────────────────

#[test]
fn flag_from_none_yields_flagged() {
    assert_eq!(next_state(None, Transition::Flag).unwrap(), State::Flagged);
}

#[test]
fn flag_from_any_non_none_state_is_illegal() {
    for s in [
        State::Flagged,
        State::Quarantining,
        State::Quarantined,
        State::Restoring,
        State::Kept,
        State::Restored,
        State::Purging,
        State::Purged,
        State::Failed,
    ] {
        assert!(
            next_state(Some(s), Transition::Flag).is_err(),
            "flag from {s:?} should be illegal"
        );
    }
}

// ── Keep / quarantine / restore / purge happy paths ──────────────────────────

#[test]
fn keep_flagged_to_kept() {
    assert_eq!(
        next_state(Some(State::Flagged), Transition::Keep).unwrap(),
        State::Kept
    );
}

#[test]
fn quarantine_flow() {
    assert_eq!(
        next_state(Some(State::Flagged), Transition::Quarantine).unwrap(),
        State::Quarantining
    );
    assert_eq!(
        next_state(Some(State::Quarantining), Transition::QuarantineComplete).unwrap(),
        State::Quarantined
    );
}

#[test]
fn restore_flow() {
    assert_eq!(
        next_state(Some(State::Quarantined), Transition::Restore).unwrap(),
        State::Restoring
    );
    assert_eq!(
        next_state(Some(State::Restoring), Transition::RestoreComplete).unwrap(),
        State::Restored
    );
}

#[test]
fn purge_flow() {
    assert_eq!(
        next_state(Some(State::Quarantined), Transition::Purge).unwrap(),
        State::Purging
    );
    assert_eq!(
        next_state(Some(State::Purging), Transition::PurgeComplete).unwrap(),
        State::Purged
    );
}

// ── Fail is only legal from in-flight states ─────────────────────────────────

#[test]
fn fail_legal_only_from_in_flight() {
    for s in [State::Quarantining, State::Restoring, State::Purging] {
        assert_eq!(
            next_state(Some(s), Transition::Fail).unwrap(),
            State::Failed,
            "fail should be legal from {s:?}"
        );
    }
    for s in [
        State::Flagged,
        State::Quarantined,
        State::Kept,
        State::Restored,
        State::Purged,
        State::Failed,
    ] {
        assert!(
            next_state(Some(s), Transition::Fail).is_err(),
            "fail should be illegal from {s:?}"
        );
    }
}

// ── Terminal states reject all transitions ───────────────────────────────────

#[test]
fn terminal_states_reject_all_transitions() {
    let terminals = [State::Kept, State::Restored, State::Purged, State::Failed];
    let transitions = [
        Transition::Keep,
        Transition::Quarantine,
        Transition::QuarantineComplete,
        Transition::Restore,
        Transition::RestoreComplete,
        Transition::Purge,
        Transition::PurgeComplete,
        Transition::Fail,
    ];
    for s in terminals {
        for t in transitions {
            assert!(
                next_state(Some(s), t).is_err(),
                "expected illegal: {s:?} --{t:?}-->"
            );
        }
    }
}

// ── No skipping the in-flight step ───────────────────────────────────────────

#[test]
fn cannot_skip_in_flight() {
    assert!(next_state(Some(State::Flagged), Transition::QuarantineComplete).is_err());
    assert!(next_state(Some(State::Quarantined), Transition::RestoreComplete).is_err());
    assert!(next_state(Some(State::Quarantined), Transition::PurgeComplete).is_err());
}

// ── Restore only from quarantined ────────────────────────────────────────────

#[test]
fn restore_only_from_quarantined() {
    assert!(next_state(Some(State::Flagged), Transition::Restore).is_err());
    assert!(next_state(Some(State::Kept), Transition::Restore).is_err());
    assert!(next_state(Some(State::Restored), Transition::Restore).is_err());
    assert!(next_state(Some(State::Failed), Transition::Restore).is_err());
    assert!(next_state(Some(State::Purged), Transition::Restore).is_err());
}

// ── Cannot re-quarantine ─────────────────────────────────────────────────────

#[test]
fn no_requarantine_from_terminal() {
    assert!(next_state(Some(State::Restored), Transition::Quarantine).is_err());
    assert!(next_state(Some(State::Kept), Transition::Quarantine).is_err());
}

// ── Genesis is reflog-only ───────────────────────────────────────────────────

#[test]
fn genesis_never_a_legal_state_transition() {
    assert!(next_state(None, Transition::Genesis).is_err());
    assert!(next_state(Some(State::Flagged), Transition::Genesis).is_err());
}

// ── Classifier helpers ───────────────────────────────────────────────────────

#[test]
fn is_terminal_classifier() {
    assert!(is_terminal(State::Kept));
    assert!(is_terminal(State::Restored));
    assert!(is_terminal(State::Purged));
    assert!(is_terminal(State::Failed));
    assert!(!is_terminal(State::Flagged));
    assert!(!is_terminal(State::Quarantined));
    assert!(!is_terminal(State::Quarantining));
}

#[test]
fn is_in_flight_classifier() {
    assert!(is_in_flight(State::Quarantining));
    assert!(is_in_flight(State::Restoring));
    assert!(is_in_flight(State::Purging));
    assert!(!is_in_flight(State::Flagged));
    assert!(!is_in_flight(State::Quarantined));
    assert!(!is_in_flight(State::Kept));
}

// ── String round-trip (used by napi-rs binding) ──────────────────────────────

#[test]
fn state_string_roundtrip() {
    for s in [
        State::Flagged,
        State::Quarantining,
        State::Quarantined,
        State::Restoring,
        State::Kept,
        State::Restored,
        State::Purging,
        State::Purged,
        State::Failed,
    ] {
        let str = s.to_string();
        let parsed: State = str.parse().unwrap();
        assert_eq!(parsed, s);
    }
}

#[test]
fn transition_string_roundtrip() {
    for t in [
        Transition::Flag,
        Transition::Quarantine,
        Transition::QuarantineComplete,
        Transition::Restore,
        Transition::RestoreComplete,
        Transition::Keep,
        Transition::Purge,
        Transition::PurgeComplete,
        Transition::Fail,
        Transition::Genesis,
    ] {
        let s = t.to_string();
        let parsed: Transition = s.parse().unwrap();
        assert_eq!(parsed, t);
    }
}

// ── Error shape ──────────────────────────────────────────────────────────────

#[test]
fn illegal_transition_carries_context() {
    let err = next_state(Some(State::Failed), Transition::Restore).unwrap_err();
    assert_eq!(err.from(), Some(State::Failed));
    assert_eq!(err.transition(), Transition::Restore);
    let msg = err.to_string();
    assert!(msg.contains("Failed"), "msg should name from state: {msg}");
    assert!(msg.contains("Restore"), "msg should name transition: {msg}");
    assert!(
        matches!(err, IllegalTransition::Invalid { .. }),
        "single variant"
    );
}
